権限不明アカウントでログイン後画面真っ白

```
react-dom.production.min.js:216 Error: Minified React error #321; visit https://reactjs.org/docs/error-decoder.html?invariant=321 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.
    at Object.oa (react-dom.production.min.js:156:258)
    at t.useContext (react.production.min.js:22:273)
    at me (index.js:764:43)
    at Ce (jotai.ts:25:24)
    at Extra.tsx:16:7
    at Dl (react-dom.production.min.js:262:359)
    at t.unstable_runWithPriority (scheduler.production.min.js:18:343)
    at Xi (react-dom.production.min.js:122:325)
    at Bl (react-dom.production.min.js:261:308)
    at xl (react-dom.production.min.js:243:50)
```

リロードすると直る

--- a/package.json
+++ b/package.json
@@ -35,8 +35,8 @@
     "workbox-streams": "^6.5.3"
   },
   "scripts": {
-    "start": "react-app-rewired start",
-    "build": "react-app-rewired build REACT_APP_VERSION=$npm_package_version",
+    "start": "react-app-rewired start REACT_APP_VERSION=\"$npm_package_version\"",
+    "build": "react-app-rewired build REACT_APP_VERSION=\"$npm_package_version\"",
     "eslint:check": "eslint src",
     "eslint": "eslint src --fix",
     "prettier:check": "prettier src --check",

--- a/src/components/block/Version.tsx
+++ b/src/components/block/Version.tsx
@@ -4,7 +4,7 @@ import { Box, Typography } from "@mui/material";
 const Version = () => {
   return (
     <Box>
-      <Typography variant="caption">ver. 1.8.6 (22/08/20)</Typography>
+      <Typography variant="caption">ver. {process.env.REACT_APP_VERSION}-{process.env.REACT_APP_ENV} (22/08/20)</Typography>
       <br />
       <Typography variant="caption">© 栄東祭実行委員会 技術部</Typography>
     </Box>

--- a/src/components/lib/commonFunction.ts
+++ b/src/components/lib/commonFunction.ts
@@ -83,7 +83,7 @@ export const handleApiError = (error: AxiosError, name: string) => {
       let content =
         "```timestamp: " + moment().format("MM/DD HH:mm:ss SSS") + "\n";
       const version = (process.env.REACT_APP_VERSION || "unknown") + "-" + (process.env.REACT_APP_ENV || "unknown");
-      content += "version   : " + version + "\n";
+      content += "version  : " + version + "\n";
       const userId = localStorage.getItem("user_id");
       if (userId) {
         content += "user_id  : " + userId + "\n";
@@ -101,6 +101,38 @@ export const handleApiError = (error: AxiosError, name: string) => {
         "\n\n" +
         String(error) +
         "```";
+      const postData = {
+        username: "error log",
+        embeds: [{ title: "REQUEST ERROR", description: content, color: parseInt("a83232", 16) }],
+      };
+      axios.post(url, postData, config).catch((err: AxiosError) => {
+        console.log(err);
+      });
+    }
+  }
+};
+
+export const sendLog = (message: string) => {
+  const env = process.env.REACT_APP_ENV;
+  if (env && (env === "production" || env === "develop")) {
+    const url = process.env.REACT_APP_DISCORD_WEBHOOK_URL;
+    if (url) {
+      const config = {
+        headers: {
+          Accept: "application/json",
+          "Content-type": "application/json",
+        },
+      };
+      let content =
+        "```timestamp: " + moment().format("MM/DD HH:mm:ss SSS") + "\n";
+      const version = (process.env.REACT_APP_VERSION || "unknown") + "-" + (process.env.REACT_APP_ENV || "unknown");
+      content += `version  : ${version}\n`;
+      const userId = localStorage.getItem("user_id");
+      if (userId) {
+        content += `user_id  : ${userId}\n`;
+      }
+      content += `location : ${window.location.pathname}\n\n${message}`;
+      content += "```";
       const postData = {
         username: "error log",
         embeds: [{ title: "ERROR", description: content, color: parseInt("a83232", 16) }],

--- a/src/components/lib/commonFunction.ts
+++ b/src/components/lib/commonFunction.ts
@@ -82,6 +82,8 @@ export const handleApiError = (error: AxiosError, name: string) => {
       };
       let content =
         "```timestamp: " + moment().format("MM/DD HH:mm:ss SSS") + "\n";
+      const version = (process.env.REACT_APP_VERSION || "unknown") + "-" + (process.env.REACT_APP_ENV || "unknown");
+      content += "version   : " + version + "\n";
       const userId = localStorage.getItem("user_id");
       if (userId) {
         content += "user_id  : " + userId + "\n";

--- a/src/components/lib/commonFunction.ts
+++ b/src/components/lib/commonFunction.ts
@@ -66,44 +66,46 @@ export const reservationIdValidation = (reservation_id: string) => {
 
 export const handleApiError = (error: AxiosError, name: string) => {
   console.log(error);
-  ReactGA.event({
-    category: `error_${name}`,
-    action: error.message,
-  });
-  const url = process.env.REACT_APP_DISCORD_WEBHOOK_URL;
-  console.log(url);
-  if (url) {
-    const config = {
-      headers: {
-        Accept: "application/json",
-        "Content-type": "application/json",
-      },
-    };
-    const userId = localStorage.getItem("user_id");
-    let content =
-      "```timestamp: " + moment().format("MM/DD HH:mm:ss SSS") + "\n";
-    if (userId) {
-      content += "user_id  : " + userId + "\n";
-    }
-    content +=
-      "type     : " +
-      name +
-      "\nlocation : " +
-      window.location.pathname;
-    if (error.code) {
-      content += "\ncode     : " + error.code;
-    }
-    content += "\nmessage  : " +
-      error.message +
-      "\n\n" +
-      String(error) +
-      "```";
-    const postData = {
-      username: "error log",
-      embeds: [{ title: "ERROR", description: content, color: parseInt("a83232", 16) }],
-    };
-    axios.post(url, postData, config).catch((err: AxiosError) => {
-      console.log(err);
+  const env = process.env.REACT_APP_ENV;
+  if (env && (env === "production" || env === "develop")) {
+    ReactGA.event({
+      category: `error_${name}`,
+      action: error.message,
     });
+    const url = process.env.REACT_APP_DISCORD_WEBHOOK_URL;
+    if (url) {
+      const config = {
+        headers: {
+          Accept: "application/json",
+          "Content-type": "application/json",
+        },
+      };
+      let content =
+        "```timestamp: " + moment().format("MM/DD HH:mm:ss SSS") + "\n";
+      const userId = localStorage.getItem("user_id");
+      if (userId) {
+        content += "user_id  : " + userId + "\n";
+      }
+      content +=
+        "type     : " +
+        name +
+        "\nlocation : " +
+        window.location.pathname;
+      if (error.code) {
+        content += "\ncode     : " + error.code;
+      }
+      content += "\nmessage  : " +
+        error.message +
+        "\n\n" +
+        String(error) +
+        "```";
+      const postData = {
+        username: "error log",
+        embeds: [{ title: "ERROR", description: content, color: parseInt("a83232", 16) }],
+      };
+      axios.post(url, postData, config).catch((err: AxiosError) => {
+        console.log(err);
+      });
+    }
   }
 };

--- a/src/components/page/Body.tsx
+++ b/src/components/page/Body.tsx
@@ -115,7 +115,7 @@ const Body = () => {
                       <Route path="exit" element={<EntranceExit />} />
                     </>
                   ) : (
-                    <Route path="*" element={<Extra type="401" />} />
+                    <Route path="*" element={<Extra type="unauthorized" />} />
                   )}
                 </Route>
                 <Route path="analytics">
@@ -136,7 +136,7 @@ const Body = () => {
                       />
                     </>
                   ) : (
-                    <Route path="*" element={<Extra type="401" />} />
+                    <Route path="*" element={<Extra type="notFound" />} />
                   )}
                 </Route>
                 <Route path="admin">
@@ -149,10 +149,10 @@ const Body = () => {
                       />
                     </>
                   ) : (
-                    <Route path="*" element={<Extra type="401" />} />
+                    <Route path="*" element={<Extra type="unauthorized" />} />
                   )}
                 </Route>
-                <Route path="*" element={<Extra type="404" />} />
+                <Route path="*" element={<Extra type="notFound" />} />
               </>
             ) : (
               <Route path="*" element={<Extra type="unknown" />} />

--- a/src/components/page/Body.tsx
+++ b/src/components/page/Body.tsx
@@ -154,6 +154,8 @@ const Body = () => {
                 </Route>
                 <Route path="*" element={<Extra type="notFound" />} />
               </>
+            ) : (profile.user_type === "keepout") ? (
+              <Route path="*" element={<Extra type="keepout" />} />
             ) : (
               <Route path="*" element={<Extra type="unknown" />} />
             )}

--- a/src/components/page/Extra.tsx
+++ b/src/components/page/Extra.tsx
@@ -5,6 +5,7 @@ import { pageTitleAtom, profileAtom, tokenAtom } from "#/components/lib/jotai";
 
 import { Card, Box, Typography, Button, Link } from "@mui/material";
 import LogoutRoundedIcon from "@mui/icons-material/LogoutRounded";
+import { sendLog } from "#/components/lib/commonFunction";
 
 type extraProp = {
   type: "notFound" | "unauthorized" | "keepout" | "unknown" | "loading";
@@ -19,6 +20,9 @@ const NotFound = (props: extraProp) => {
   useEffect(() => {
     if (props.type !== "loading") {
       setPageTitle("エラー");
+    }
+    if (props.type === "unknown") {
+      sendLog("unknown account");
     }
   }, []);
 

--- a/src/components/page/Extra.tsx
+++ b/src/components/page/Extra.tsx
@@ -1,59 +1,75 @@
 import React, { useEffect } from "react";
 import { useNavigate } from "react-router-dom";
 import { useSetAtom } from "jotai";
-import { pageTitleAtom } from "#/components/lib/jotai";
+import { pageTitleAtom, profileAtom, tokenAtom } from "#/components/lib/jotai";
 
-import { Grid, Card, Box, Typography, Button } from "@mui/material";
+import { Card, Box, Typography, Button } from "@mui/material";
+import LogoutRoundedIcon from "@mui/icons-material/LogoutRounded";
 
 type extraProp = {
   type: "404" | "401" | "unknown" | "loading";
 };
 
 const NotFound = (props: extraProp) => {
-  const navigate = useNavigate();
   const setPageTitle = useSetAtom(pageTitleAtom);
+  const navigate = useNavigate();
+  const setProfile = useSetAtom(profileAtom);
+  const setToken = useSetAtom(tokenAtom);
 
   useEffect(() => {
     if (props.type !== "loading") {
       setPageTitle("エラー");
     }
   }, []);
+
+  const logout = () => {
+    setToken(null);
+    setProfile(null);
+    localStorage.removeItem("gatewayApiToken");
+    navigate("/login", { replace: true });
+  };
+
   return (
-    <Grid container spacing={2}>
-      <Grid item xs={12}>
-        <Card variant="outlined" sx={{ p: 2, height: "100%" }}>
-          {props.type === "404" ? (
-            <Typography>お探しのページは見つかりませんでした。</Typography>
-          ) : props.type === "401" ? (
-            <Typography>このページを表示する権限がありません。</Typography>
-          ) : props.type === "unknown" ? (
-            <>
-              <Typography>
-                このアカウントは使えません。テスト用のアカウントである可能性があります。
-              </Typography>
-              <Typography>
-                団体の代表者へメールでアカウントを配布しています。自分の団体に配られているアカウントを使用してください。
-              </Typography>
-              <Typography>
-                メールを受け取っていない場合は技術部までお問い合わせください。
-              </Typography>
-            </>
-          ) : props.type === "loading" ? (
-            <Typography>ログインセッションの検証中...</Typography>
-          ) : (
-            <Typography>何らかのエラーが発生しました。</Typography>
-          )}
-          <Box sx={{ textAlign: "right" }}>
-            <Button
-              onClick={() => navigate("/", { replace: true })}
-              variant="outlined"
-            >
-              トップに戻る
-            </Button>
-          </Box>
-        </Card>
-      </Grid>
-    </Grid>
+    <Card variant="outlined" sx={{ p: 2, height: "100%" }}>
+      {props.type === "404" ? (
+        <Typography>お探しのページは見つかりませんでした。</Typography>
+      ) : props.type === "401" ? (
+        <Typography>このページを表示する権限がありません。</Typography>
+      ) : props.type === "unknown" ? (
+        <>
+          <Typography>
+            このアカウントは使えません。テスト用のアカウントである可能性があります。
+          </Typography>
+          <Typography>
+            団体の代表者へメールでアカウントを配布しています。自分の団体に配られているアカウントを使用してください。
+          </Typography>
+          <Typography>
+            メールを受け取っていない場合は技術部までお問い合わせください。
+          </Typography>
+        </>
+      ) : props.type === "loading" ? (
+        <Typography>ログインセッションの検証中...</Typography>
+      ) : (
+        <Typography>何らかのエラーが発生しました。</Typography>
+      )}
+      <Box sx={{ mt: 2, textAlign: "right" }}>
+        <Button
+          variant="outlined"
+          color="error"
+          onClick={logout}
+          startIcon={<LogoutRoundedIcon />}
+          sx={{ mr: 2 }}
+        >
+          ログアウト
+        </Button>
+        <Button
+          onClick={() => navigate("/", { replace: true })}
+          variant="outlined"
+        >
+          トップに戻る
+        </Button>
+      </Box>
+    </Card>
   );
 };
 

--- a/src/components/page/Extra.tsx
+++ b/src/components/page/Extra.tsx
@@ -3,11 +3,11 @@ import { useNavigate } from "react-router-dom";
 import { useSetAtom } from "jotai";
 import { pageTitleAtom, profileAtom, tokenAtom } from "#/components/lib/jotai";
 
-import { Card, Box, Typography, Button } from "@mui/material";
+import { Card, Box, Typography, Button, Link } from "@mui/material";
 import LogoutRoundedIcon from "@mui/icons-material/LogoutRounded";
 
 type extraProp = {
-  type: "notFound" | "unauthorized" | "unknown" | "loading";
+  type: "notFound" | "unauthorized" | "keepout" | "unknown" | "loading";
 };
 
 const NotFound = (props: extraProp) => {
@@ -35,6 +35,16 @@ const NotFound = (props: extraProp) => {
         <Typography>お探しのページは見つかりませんでした。</Typography>
       ) : props.type === "unauthorized" ? (
         <Typography>このページを表示する権限がありません。</Typography>
+      ) : props.type === "keepout" ? (
+        <Typography>
+          現在メンテナンス中です。最新の情報は<Link
+            href={process.env.REACT_APP_MANUAL_URL || "/"}
+            target="_blank"
+            underline="hover"
+          >
+            こちら
+          </Link>からご確認ください。
+        </Typography>
       ) : props.type === "unknown" ? (
         <>
           <Typography>

--- a/src/components/page/Extra.tsx
+++ b/src/components/page/Extra.tsx
@@ -7,7 +7,7 @@ import { Card, Box, Typography, Button } from "@mui/material";
 import LogoutRoundedIcon from "@mui/icons-material/LogoutRounded";
 
 type extraProp = {
-  type: "404" | "401" | "unknown" | "loading";
+  type: "notFound" | "unauthorized" | "unknown" | "loading";
 };
 
 const NotFound = (props: extraProp) => {
@@ -31,9 +31,9 @@ const NotFound = (props: extraProp) => {
 
   return (
     <Card variant="outlined" sx={{ p: 2, height: "100%" }}>
-      {props.type === "404" ? (
+      {props.type === "notFound" ? (
         <Typography>お探しのページは見つかりませんでした。</Typography>
-      ) : props.type === "401" ? (
+      ) : props.type === "unauthorized" ? (
         <Typography>このページを表示する権限がありません。</Typography>
       ) : props.type === "unknown" ? (
         <>

--- a/src/components/page/Extra.tsx
+++ b/src/components/page/Extra.tsx
@@ -1,6 +1,7 @@
 import React, { useEffect } from "react";
 import { useNavigate } from "react-router-dom";
-import { setTitle } from "#/components/lib/jotai";
+import { useSetAtom } from "jotai";
+import { pageTitleAtom } from "#/components/lib/jotai";
 
 import { Grid, Card, Box, Typography, Button } from "@mui/material";
 
@@ -10,10 +11,11 @@ type extraProp = {
 
 const NotFound = (props: extraProp) => {
   const navigate = useNavigate();
+  const setPageTitle = useSetAtom(pageTitleAtom);
 
   useEffect(() => {
     if (props.type !== "loading") {
-      setTitle("エラー");
+      setPageTitle("エラー");
     }
   }, []);
   return (

--- a/src/components/page/Login.tsx
+++ b/src/components/page/Login.tsx
@@ -80,6 +80,7 @@ const Login = () => {
         setLoading(false);
       });
   };
+
   return (
     <Grid container spacing={2}>
       <Grid item xs={12}>

--- a/src/components/page/Login.tsx
+++ b/src/components/page/Login.tsx
@@ -1,6 +1,6 @@
-import React, { useState } from "react";
+import React, { useEffect, useState } from "react";
 import { useNavigate } from "react-router-dom";
-import { useSetAtom } from "jotai";
+import { useAtom, useSetAtom } from "jotai";
 import { tokenAtom, profileAtom, setTitle } from "#/components/lib/jotai";
 import { AxiosError } from "axios";
 import apiClient from "#/axios-config";
@@ -25,12 +25,18 @@ import { handleApiError } from "#/components/lib/commonFunction";
 const Login = () => {
   setTitle("ログイン");
   const setToken = useSetAtom(tokenAtom);
-  const setProfile = useSetAtom(profileAtom);
+  const [profile, setProfile] = useAtom(profileAtom);
   const navigate = useNavigate();
   const [loading, setLoading] = useState<boolean>(false);
   const [userIdValue, setUserIdValue] = useState<string>("");
   const [passwordValue, setPasswordValue] = useState<string>("");
   const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  useEffect(() => {
+    if (profile) {
+      navigate("/", { replace: true });
+    }
+  }, [profile]);
 
   const login = () => {
     localStorage.setItem("user_id", userIdValue);
@@ -52,13 +58,12 @@ const Login = () => {
             },
           })
           .then((meRes) => {
-            setProfile(meRes);
-            navigate("/", { replace: true });
             ReactGA.event({
               category: "login",
               action: "success",
               label: userIdValue,
             });
+            setProfile(meRes);
           })
           .catch((err: AxiosError) => {
             handleApiError(err, "get_user_info");

